designs/retries.md

```diff
@@ -0,0 +1,170 @@
+# Retries
+
+Operation requests might fail for a number of reasons that are unrelated to the
+input paramters, such as a transient network issue, or excessive load on the
+service. This document describes how Smithy clients will automatically retry in
+those cases, and how the retry system can be modified.
+
+## Specification
+
+Retry behavior will be determined by a `RetryStrategy`. Implementations of the
+`RetryStrategy` will produce `RetryToken`s that carry metadata about the
+invocation, notably the number of attempts that have occurred and the amount of
+time that must pass before the next attempt. Passing state through tokens in
+this way allows the `RetryStrategy` itself to be isolated from the state of an
+individual request.
+
+```python
+@dataclass(kw_only=True)
+class RetryToken(Protocol):
+    retry_count: int
+    """Retry count is the total number of attempts minus the initial attempt."""
+
+    retry_delay: float
+    """Delay in seconds to wait before the retry attempt."""
+
+
+class RetryStrategy(Protocol):
+    backoff_strategy: RetryBackoffStrategy
+    """The strategy used by returned tokens to compute delay duration values."""
+
+    max_attempts: int
+    """Upper limit on total attempt count (initial attempt plus retries)."""
+
+    def acquire_initial_retry_token(
+        self, *, token_scope: str | None = None
+    ) -> RetryToken:
+        """Called before any retries (for the first attempt at the operation).
+
+        :param token_scope: An arbitrary string accepted by the retry strategy to
+            separate tokens into scopes.
+        :returns: A retry token, to be used for determining the retry delay, refreshing
+            the token after a failure, and recording success after success.
+        :raises RetryError: If the retry strategy has no available tokens.
+        """
+        ...
+
+    def refresh_retry_token_for_retry(
+        self, *, token_to_renew: RetryToken, error: Exception
+    ) -> RetryToken:
+        """Replace an existing retry token from a failed attempt with a new token.
+
+        :param token_to_renew: The token used for the previous failed attempt.
+        :param error: The error that triggered the need for a retry.
+        :raises RetryError: If no further retry attempts are allowed.
+        """
+        ...
+
+    def record_success(self, *, token: RetryToken) -> None:
+        """Return token after successful completion of an operation.
+
+        :param token: The token used for the previous successful attempt.
+        """
+        ...
+```
+
+A request using a `RetryStrategy` would look something like the following
+example:
+
+```python
```

Comment:
Can you break down this example to help me understand it better? I find it really confusing

Reply:
Update: The content below helped by understand this example better:
> RetryStrategy implementations MUST raise a RetryError if they receive an exception where is_retry_safe is False and SHOULD raise a RetryError if it is None. RetryStrategy implementations SHOULD use a delay that is at least as long as retry_after but MAY choose to wait longer.

Reply:
I moved the example to the end to avoid this confusion